### ISSUE TITLE
Similar to disk.frame, should we just collaborate?

### ISSUE THREAD

**Issue by xiaodaigh** [krlmlr/fstplyr#10]
See https://github.com/xiaodaigh/disk.frame

**Comment by MarcusKlik** [comment]
Hi @xiaodaigh, your package looks very nice and promising for working with large on-disk datasets efficiently. However, I think that the approaches of `fstplyr` and `disk.frame` are a little different. From what I can gather (correct me if I'm wrong):

* `fst` was currently chosen as the backend for operations, but your package is not bound to `fst`. Your framework can in principle be used on top of other backends as well (it's not `fst` specific).
* All operations are chunk based and the verbs work on each chunk separately. That means that output might be different from what users expect from `dplyr` verbs:

``` r
library(disk.frame)
library(fst)

rows_per_chunk <- 1000
nr_of_chunks <- 8

# write chunks
dir.create("tmp")
chunks <- lapply(1:nr_of_chunks, function(x) {
  chunk_data <- data.frame(a = runif(rows_per_chunk), b = runif(rows_per_chunk))
  write.fst(chunk_data, paste0("tmp/", x, ".fst"))
})

df = disk.frame("tmp")

# some selection and filtering
df2 <- df %>%
  select(a) %>%
  filter(a < 0.5)

# show modified chunk
head(disk.frame::get_chunk(df2, 3))
#>             a
#> 1: 0.18581499
#> 2: 0.44015172
#> 3: 0.11832342
#> 4: 0.41180058
#> 5: 0.07036826
#> 6: 0.27242632
```

So your package provides a framework for processing many chunks in parallel, performing the same operations on each of them, is that correct? And grouping is also done on a per-chunk basis?

I think `fstplyr`'s approach is much simpler. It's a (`fst` only) `dplyr` backend for batches of `fst` files but a single dataset still maps to a single `fst` file, a 1-to-1 relation between the dataset and stored data.

So similar but also very different :-).

**Comment by xiaodaigh** [comment]
@MarcusKlik Yes grouping is done on a per-chunk basis, however, if the group-by variables with the same values are in the same chunk then `group_by.disk.frame` essentially produces the same result as `group_by`. You can force this using the hard  = TRUE argument e.g. `group_by(by_var, hard= TRUE)`.

Ok, I don't mind re-using `fstplyr`'s filtering mechanism and select mechanism here, but yeah I guess they are quite different. But `disk.frame` could use something like `fstplyr` as the backend for dplyr verbs is what I am thinking about now.

**Comment by MarcusKlik** [comment]
Hi @xiaodaigh, yes, that's a nice idea and in that case the package dependencies would be clear. Also, you could support multiple `?plyr` backends and that would make your package very flexible 👍
